src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs

```diff
@@ -0,0 +1,261 @@
+        /// </summary>
+        public bool IsReadOnly => false;
+
+        /// <summary>
+        /// Gets the number of elements contained in the collection.
+        /// </summary>
+        public int Count => _list.Count;
+
+        /// <summary>
+        /// Adds an tag with the provided key and value to the collection.
+        /// This collection doesn't allow adding two tags with the same key.
+        /// </summary>
+        /// <param name="key">The tag key.</param>
+        /// <param name="value">The tag value.</param>
+        public void Add(string key, object value)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            int index = _list.FindIndex(kvp => kvp.Key == key);
+            if (index >= 0)
+            {
+                throw new InvalidOperationException(SR.Format(SR.KeyAlreadyExist, key));
+            }
+
+            _list.Add(new KeyValuePair<string, object>(key, value));
+        }
+
+        /// <summary>
+        /// Adds an item to the collection
+        /// </summary>
+        /// <param name="item">Key and value pair of the tag to add to the collection.</param>
+        public void Add(KeyValuePair<string, object> item)
+        {
+            if (item.Key == null)
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+
+            int index = _list.FindIndex(kvp => kvp.Key == item.Key);
+            if (index >= 0)
+            {
+                throw new InvalidOperationException(SR.Format(SR.KeyAlreadyExist, item.Key));
+            }
+
+            _list.Add(item);
+        }
+
+        /// <summary>
+        /// Removes all items from the collection.
+        /// </summary>
+        public void Clear() => _list.Clear();
+
+        public bool Contains(KeyValuePair<string, object> item) => _list.Contains(item);
+
+        /// <summary>
+        /// Determines whether the collection contains an element with the specified key.
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns>True if the collection contains tag with that key. False otherwise.</returns>
+        public bool ContainsKey(string key) => _list.FindIndex(kvp => kvp.Key == key) >= 0;
+
+        /// <summary>
+        /// Copies the elements of the collection to an array, starting at a particular array index.
+        /// </summary>
+        /// <param name="array">The array that is the destination of the elements copied from collection.</param>
+        /// <param name="arrayIndex">The zero-based index in array at which copying begins.</param>
+        public void CopyTo(KeyValuePair<string, object>[] array, int arrayIndex) => _list.CopyTo(array, arrayIndex);
+
+        /// <summary>
+        /// Returns an enumerator that iterates through the collection.
+        /// </summary>
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator() => _list.GetEnumerator();
```

Comment:
Possible to add a `struct` enumerator here? It will be common in OpenTelemetry exporter scenario to `foreach` over the tags while building outgoing messages. `struct` enumerator would avoid an allocation when we do that.

Reply:
_list.GetEnumerator will return struct already. https://source.dot.net/#System.Private.CoreLib/List.cs,1075

Reply:
Yes exactly. We need something similar. The new collection only exposes:
```csharp
        public System.Collections.Generic.IEnumerator<System.Collections.Generic.KeyValuePair<string, object>> GetEnumerator() { throw null; }
        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw null; }
```
So any `foreach` over ActivityTagsCollection will have to box that `struct List<T>.Enumerator`. ~20 byte allocation on 64 bit, last time I looked at it.

Reply:
We have this code in the OT library that will look for a `GetEnumerator` method that returns `struct` and builds a DynamicMethod to use that for enumerations: https://github.com/open-telemetry/opentelemetry-dotnet/blob/3214f1cd963f8e34dbdcddcc3ae923191a8f0a11/src/OpenTelemetry/Internal/EnumerationHelper.cs#L96

We could just expose the `List<T>.Enumerator` that would work I think:
`public System.Collections.Generic.List<System.Collections.Generic.KeyValuePair<string, object>>.Enumerator GetEnumerator() { throw null; }`

Not sure if co-opting a type like that will fly though.

Reply:
Thanks again. I wouldn't tie the exposed interfaces to List<T> to give us the flexibility to change the underlying implementation if we need to anytime. I'll look at exposing ActivityTagsCollection.Enumerator and return will have GetEnumerator returning this struct.

Reply:
Great thanks @tarekgh.

Reply:
I have applied the suggestion here.